### DISCUSSION TITLE
[workflow] Injecting all the workflows into a bundle service class

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#47983] [Q&A] [answered]
Now that injecting the workflow registry is deprecated, what is the proper way to inject all the workflows into a service.

```php
      
class WorkflowHelperService
{
    public function __construct(
        private ServiceLocator $locator,
```

and in my bundle config
```php
    public function loadExtension(array $config, ContainerConfigurator $container, ContainerBuilder $builder): void
    {
        $builder->autowire( WorkflowHelperService::class)
            ->setAutoconfigured(true)
        ;
     }}}


    public function build(ContainerBuilder $container): void
    {
        parent::build($container);
        $container->addCompilerPass($this);
    }
    // The compiler pass
    public function process(ContainerBuilder $container): void
    {
        $locateableServices = ['workflow' => new Reference('workflow')];

        $workflowHelperService = $container->get(WorkflowHelperService::class);
        $workflowHelperService->setArgument('$locator', ServiceLocatorTagPass::register($container, $locateableServices));
    }
```

This throws an error: 

```
Constructing service "state_machine.Task" from a parent definition is not supported at build time.
```

I imagine that I'm not getting the services correctly in the compiler pass, I'm trying to follow https://symfony.com/doc/current/service_container/service_subscribers_locators.html#using-service-locators-in-compiler-passes but obviously not doing something right.  Thanks.

**Comment by MatTheCat** [comment]
In compiler passes you don’t interact with services but with their definition. Plus you don’t have to build the locator yourself as the services you want already are tagged.

```php
    public function process(ContainerBuilder $container): void
    {
        $workflowHelperDefinition = $container->findDefinition(WorkflowHelperService::class);
        $workflowHelperDefinition->setArgument('$locator', tagged_locator('workflow', 'name'));
    }
```

**Comment by tacman** [comment]
Hmm, I'm still not getting anything:

```php
    public function process(ContainerBuilder $container): void
    {

        dump(tagged_locator('workflow.state_machine', 'name')->getValues());
        dump(tagged_locator('workflow, 'name')->getValues());

    }
```

both return empty arrays, but when I get the workflow from the $registry, they're there.

**Reply by MatTheCat** [reply]
Please write in “Write a reply” instead of “Suggest an answer” to reply to a message.

What are you expecting by dumping a tagged locator values? These are not known until the container is compiled.
Did you try to replace your `process` method by the one in https://github.com/symfony/symfony/discussions/47983#discussioncomment-3993170? Didn’t it work?

**Reply by tacman** [reply]
I'm trying to get a list of all the current workflows.

In my current code, I inject the registry, then get the workflows via the none-too-pretty code:

```php
        $registry = $this->workflowRegistry;
        $reflectionProperty = new \ReflectionProperty(get_class($this->workflowRegistry), 'workflows');
        $workflowBlobs = $reflectionProperty->getValue($this->workflowRegistry);
``` 

Since injecting the registry is deprecated, @lyrixx mentions https://github.com/symfony/symfony/pull/46000 that the workflows can now be accessed by tagged_locator:

```
It can be legit to need all workflows, for documentation or for testing.
So, all workflow services are tagged and can be injected with the
following YAML syntax:

!tagged_locator  { tag: workflow, index_by: name }
or PHP syntax:

tagged_locator('workflow', 'name')
```

BUT it's not working for me, in the same way that injecting the registry and reflecting back the workflows does.

**Reply by MatTheCat** [reply]
If you want to be able to list the workflows you need a tagged iterator, not a tagged locator.

**Reply by tacman** [reply]
Thanks.  Can you show me how to do that?  Is is the compiler pass the right place?   I still find that part of Symfony mystifying.

Replacing tagged_locatoor with tagged_iterator in my code didn't change anything, the result is still empty.

**Reply by MatTheCat** [reply]
A service _locator_ is an instance of [`ContainerInterface`](https://github.com/php-fig/container/blob/master/src/ContainerInterface.php); you can see it only exposes a `has` and `get` method which mean you have to know your services index to get them. A _tagged_ locator is just a simpler way to build a service locator using service tags.

A tagged _iterator_ will expose services through an `iterable` so that means you can iterate over to get them all. This is apparently what you want, so your `process` method should be

```php
public function process(ContainerBuilder $container): void
{
    $workflowHelperDefinition = $container->findDefinition(WorkflowHelperService::class);
    $workflowHelperDefinition->setArgument('$workflows', tagged_iterator('workflow', 'name'));
}
```

And in your `WorkflowHelperService`’s:

```php

/** @var WorkflowInterface[] */
private array $workflows = [];

public function __construct(iterable $workflows)
{
    foreach ($workflows as $name => $workflow) {
        // You can dump($workflow); here
        $this->workflows[$name] = $workflow;
    }
}
```

**Reply by tacman** [reply]
Indeed, that's have I've coded it, and it's returning 0 workflows.

```php
    // The compiler pass in SurvosWorkflowBundle
    public function process(ContainerBuilder $container): void
    {
        $workflowHelperDefinition = $container->findDefinition(WorkflowHelperService::class);
        // the tagged iterator returns an empty result set
        $workflowHelperDefinition->setArgument('$workflows', tagged_iterator('workflow', 'name'));
    }

    public function loadExtension(array $config, ContainerConfigurator $container, ContainerBuilder $builder): void
    {
        // injecting the registry gives access to the workflows, the setArgument is there for clarity, but unnecessary because of autoConfigure
        $builder->autowire( WorkflowHelperService::class)
            ->setArgument('$workflowRegistry', new Reference('workflow.registry'))
            ->setAutoconfigured(true);
       }
}


```

```php

class WorkflowHelperService
{
    public function __construct(
        /** @var WorkflowInterface[] */
        private iterable $workflows,
        private ?Registry $workflowRegistry = null,
    ) {
        // get the workflows from the registry:
        $reflectionProperty = new \ReflectionProperty(get_class($this->workflowRegistry), 'workflows');
        $workflowCountFromRegistry = count($reflectionProperty->getValue($this->workflowRegistry));
        assert(count($this->workflows) == $workflowCountFromRegistry, sprintf("$workflowCountFromRegistry workflows in registry, %d from the service_iterator",
}

**Reply by MatTheCat** [reply]
This has nothing to do with the code I told you to use.

**Reply by tacman** [reply]
I'm not sure how you can say that, I'm literally cutting and pasting from your suggestions.  Yeah, injected the registry and I put an assert in there to demonstrate that there are workflows, but I'll get rid of those.

Here it is again, which just a dump, after cutting and pasting.  When you say "nothing to do with the code", can you help me out with what you mean?  

```php
class WorkflowHelperService
{
    /** @var WorkflowInterface[] */
    private array $workflows = [];

    public function __construct(iterable $workflows)
    {
        foreach ($workflows as $name => $workflow) {
            // You can dump($workflow); here
            dd($workflow);
            $this->workflows[$name] = $workflow;
        }
        assert(false, "we should only get here if there are no workflows");
    }
```

```php

    public function process(ContainerBuilder $container): void
    {
        $workflowHelperDefinition = $container->findDefinition(WorkflowHelperService::class);
        $workflowHelperDefinition->setArgument('$workflows', tagged_iterator('workflow', 'name'));
    }

    public function loadExtension(array $config, ContainerConfigurator $container, ContainerBuilder $builder): void
    {

        $builder->autowire( WorkflowHelperService::class)
            ->setAutoconfigured(true);
```

**Reply by MatTheCat** [reply]
I don’t know how you can have so many differences by copy-pasting. Do you really want me to generate a diff between

```php
/** @var WorkflowInterface[] */
private array $workflows = [];

public function __construct(iterable $workflows)
{
    foreach ($workflows as $name => $workflow) {
        // You can dump($workflow); here
        $this->workflows[$name] = $workflow;
    }
}
```
and
```php
public function __construct(
        /** @var WorkflowInterface[] */
        private iterable $workflows,
        private ?Registry $workflowRegistry = null,
    ) {
        // get the workflows from the registry:
        $reflectionProperty = new \ReflectionProperty(get_class($this->workflowRegistry), 'workflows');
        $workflowCountFromRegistry = count($reflectionProperty->getValue($this->workflowRegistry));
        assert(count($this->workflows) == $workflowCountFromRegistry, sprintf("$workflowCountFromRegistry workflows in registry, %d from the service_iterator",
}
```
?

I’ll be backing off now because it don’t seem you have the level to understand what you do, so it wouldn’t help that I do it for you.

Really, try to understand what you do. Good luck!

**Reply by tacman** [reply]
In my most recent comment, the only difference is the dump.  And I put that in, because without it, it's hard to show the problem.

Yes, in my earlier comments I used a format that skipped making $workflows private outside of the constructor.

I am cutting and pasting.

Yours:

![image](https://user-images.githubusercontent.com/619585/198825863-eff55107-ad02-419a-907a-dc43fd7f7a51.png)

Mine:

![image](https://user-images.githubusercontent.com/619585/198825892-37501cd9-9ff7-4768-8fac-49cc2240cb2d.png)

I appreciate the help, really.

**Reply by tacman** [reply]
My gut feeling is that something is out of order, my bundle is being loaded before the workflows are processed, and I need to defer the compiler pass, or something like that. And yes, I don't understand the intricacies tagged services and how/when to load them when they're related. The only reason I put in the reflection class and assertion was to demonstrate that injecting the registry works (the workflow are there), but injecting the tagged_iterator doesn't. But it appears that I came off as lacking understanding by doing that, which is my in my last comment I literally cut and pasted your code and added a dd/assert.

**Comment by tacman** [comment]
@lyrixx any chance you can take a look at this?  I'm stuck on how to access all the workflows from within my bundle.  I was trying to be clear by showing that I _could_ access it one way but not the way @MatTheCat suggested, but that backfired.

Thanks.

**Reply by lyrixx** [reply]
Can you create a small reproducer? Thanks

**Reply by tacman** [reply]
In the end, I think the problem is that I was testing for the workflows too early -- further inspection revealed that I appeared to be checking a generator, not at the actual services.  I could be wrong about where my logic error was, but when I actually iterate through the tagged_iterator response, I'm getting back the workflow services I expected.  

I am stuck, though, on how to get the support/supportStrategy from a workflow.  That is, given a workflow, which classes does it support?  Or which classes support workflows?

Previously, I got the classes through a reflection call to the now-deprecated registry:

```php

        $reflectionProperty = new \ReflectionProperty(get_class($this->workflowRegistry), 'workflows');
        $workflowBlobs = $reflectionProperty->getValue($this->workflowRegistry);
        foreach ($workflowBlobs as [$stateMachine, $suportStrategy]) {
            $class = $suportStrategy->getClassName();
```

Perhaps there's some way to query for classes implementing WorkflowSupportStrategyInterface?  

Thanks.

**Reply by lyrixx** [reply]
With a reproducer, I'll be easier to help you 

Maybe this repo can help you 
https://github.com/lyrixx/SFLive-Paris2016-Workflow

**Reply by tacman** [reply]
OK, my fork and branch now shows the issue:

```bash
git clone git@github.com:tacman/SFLive-Paris2016-Workflow.git workflow-demo && cd workflow-demo && git checkout symfony-6.1 && composer install && symfony server:start
```

Open the site and you'll see:

![image](https://user-images.githubusercontent.com/619585/200869006-0b86b435-36b7-42ef-a24e-bc87eaa07fa8.png)

My question: How can I get the classes a workflow supports and the workflows that a class supports?  I can get that information when injecting the registry, but how can I get that information without injecting the registry.

Thanks.

**Reply by lyrixx** [reply]
Thank you very much for the reproducer and the one liner command to open the project. This is very appreciated! It works like a charm.

I understand what you want to achieve. Just to be sure: You want a way to retrieve the raw configuration of all workflows, and be able to display some data for them, isn't?

I saw the code in `vendor/survos/workflow-helper-bundle/src/Service/WorkflowHelperService.php` and you used to use the reflection to achieve this, and this not a good way to do that, since you relies on private (/internal) properties, so it may brokes at anytime.

I'll try to find a better alternative ASAP.

**Reply by tacman** [reply]
No doubt, the reflection is a hack from the pit of hell.  I can't wait to get rid of it!

But yes, I want access to everything there is to know about the workflow, including what classes it supports.  The idea is that my helper bundle can hide the implementation details and allow developers (and their clients) to have an interactive playground to see the workflows, and also to generate documentation.

**Reply by tacman** [reply]
I'm so excited you're looking at this!  I first brought it up in 2016, it was marked as wont-fix, though eventually most of the definition was made available: https://github.com/symfony/symfony/issues/19938

**Comment by lyrixx** [comment, marked as answer]
Here we go.

First, with the @MatTheCat's
[answer](https://github.com/symfony/symfony/discussions/47983#discussioncomment-4002485)
you are able to get all workflows. I'll expand a bit the answer:

```yaml
services:
    _defaults:
        autowire: true
        autoconfigure: true
        bind:
            Psr\Container\ContainerInterface $workflows: !tagged_locator { tag: workflow, index_by: name } # first option
            iterable $workflows: !tagged_iterator { tag: workflow } # second options
```

With the first option, you'll be able to get workflow thanks to its name defined
in the config.

```php
class HomepageController extends AbstractController
{
    public function __construct(
        private ContainerInterface $workflows,
    ) {
    }

    #[Route('/', name: 'homepage')]
    public function index()
    {
        dd($this->workflows->get('article'));
```

With the second option, you'll get an iterator:

```
class HomepageController extends AbstractController
{
    public function __construct(
        private iterable $workflows,
    ) {
    }

    #[Route('/', name: 'homepage')]
    public function index()
    {
        foreach ($this->workflows as $workflow) {
            dd($workflow);
        }
```

Then with a workflow instance, you are able to get its definition, and all
informations you may need. Everything is public. BTW, I already use all theses
information on the [demo](https://github.com/lyrixx/SFLive-Paris2016-Workflow).
I let you play with the [online
version](https://symfony-workflow-demo.herokuapp.com/article/show/4051).

```
$definition = $workflow->getDefinition();
dd($definition);
```

![image](https://user-images.githubusercontent.com/408368/201173519-7dd0bc6c-fa13-4f7b-888d-61953246724f.png)

---

So now, you want an extra information: the relation between a workflow and its
class. IMHO this information is **internal** and **should not be displayed** to
the end user. It does not bring anything useful. More over, this "connection"
was present in the very first version only to simplify workflow usage from Twig.

But, for debugging purpose, why not :) So I'll give you a way to get this
information, it clean and stable I guess, **but you should use it with care!**
You can use a compiler pass to get the framework bundle configuration. The
configuration is already merged / normalized.

```
    public function process(ContainerBuilder $container)
    {
        $configs = $container->getExtensionConfig('framework');

        $configuration = $container
            ->getExtension('framework')
            ->getConfiguration($configs, $container)
        ;

        $config = (new Processor())->processConfiguration($configuration, $configs);

        $container->setParameter('workflows.configuration', $config['workflows']['workflows'] ?? []);
    }
```

Then use the parameter where you need it.

I [pushed a branch on my repo](https://github.com/lyrixx/SFLive-Paris2016-Workflow/commit/01f4428500415e331c43c4ad88093f8fd0d874df), on top of your work. But I'll drop it at some point.

I hope I answered all your questions. Feel free to reply under answer if you
need more information.

**Reply by tacman** [reply]
THANKS!!!

I'm refactoring my bundle now accordingly, Much appreciated.  

And with this new-found knowledge of the compiler pass, I'll be able to play around with my idea of using attributes on constants to define the workflow: https://github.com/symfony/symfony/discussions/47899

Again, muchas gracias.

**Reply by tacman** [reply]
After diving into this a bit, I remembered why I needed the class: to get the workflow metadata from a twig call.

Question: why does workflow_metadata twig function require an object?    {{ workflow_metadata(blog_post, 'max_num_of_words', place) ?

Does it really return the metadata for a specific subject, as the comment says?  It seems like it's only using the subject to find the workflow, then returns the workflow metadata, not metadata related the specific subject.

```php
    /**
     * Returns the metadata for a specific subject.
     *
     * @param string|Transition|null $metadataSubject Use null to get workflow metadata
     *                                                Use a string (the place name) to get place metadata
     *                                                Use a Transition instance to get transition metadata
     */
    public function getMetadata(object $subject, string $key, string|Transition $metadataSubject = null, string $name = null)
    {
        return $this
            ->workflowRegistry
            ->get($subject, $name)
            ->getMetadataStore()
            ->getMetadata($key, $metadataSubject)
        ;
    }
```

Again, maybe I'm misunderstanding something, but (brainstorming), suppose the subject were allowed to be null, then it simply returned the metadata for the (obviously required) workflow $name?

I realized that a lot of my code has to do with generating a fake entity just so I can get the workflow metadata.

**Reply by lyrixx** [reply]
Yes

**Comment by tacman** [comment]
So I'm back to the problem that I had earlier -- the workflow iterator doesn't return any workflows _in some situations_.  Alas, I can't figure out what that is, which is why I had all these "assert" tests earlier in this thread.

I think it has to do with the order than something is loaded.  Curiously, the framework config is correct, it returns the workflow array data.  BUT the tagged_iterator doesn't.

I am trying to reproduce this behavior, it's working on my fork of the demo but not another project.

**Comment by tacman** [comment]
I'm mystified.  Two apps, using the same bundle, with nearly identical code, but in one the tagged_iterator does not return any workflows.

As before, this works, injects the configuration into my workflow-helper bundle instead of into the container.

```bash
git clone git@github.com:tacman/SFLive-Paris2016-Workflow.git workflow-demo && cd workflow-demo && git checkout symfony-6.1 && composer install && symfony server:start
```

A nearly identical app does not work, but the config is injected, just not the workflows:

```bash
git clone git@github.com:survos/workflow-bundle-demo.git broken-workflow-demo && cd broken-workflow-demo  && composer install && yarn install && yarn dev && symfony server:start
```

![image](https://user-images.githubusercontent.com/619585/201363809-97c7d089-8c35-4e8b-b01d-454788344794.png)

I'm going crazy trying to figure out what could possibly differ.

**Reply by tacman** [reply]
Is there a way I can set the priority of my bundle?  Gut feeling is still that something is out of order.

**Comment by tacman** [comment]
At long last, I've figured it out.  Using Symfony 6.1, injecting the tagged_iterator doesn't return the workflows.

```bash
git clone git@github.com:tacman/SFLive-Paris2016-Workflow.git workflow-demo && cd workflow-demo && git checkout symfony-6.1 && composer install && symfony server:start
```

![image](https://user-images.githubusercontent.com/619585/201375912-14508048-5784-4e15-81e0-843c7c58e0ad.png)

^c the CLI server, and bump to 6.2

```bash
composer config minimum-stability beta && composer update
symfony server:start
```

![image](https://user-images.githubusercontent.com/619585/201376928-49280e8e-3903-47bc-823f-17452d87d5ec.png)

So I guess injecting by tagged_iterator is new in Symfony 6.2 for the workflow component?

**Reply by lyrixx** [reply]
Yes
